File: GuardianMonsters/core/src/de/limbusdev/guardianmonsters/fwmengine/battle/model/BattleResult.java
package de.limbusdev.guardianmonsters.fwmengine.battle.model;

import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.ArrayMap;

import de.limbusdev.guardianmonsters.guardians.items.Item;
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian;
import de.limbusdev.guardianmonsters.guardians.monsters.Guardian;
import de.limbusdev.guardianmonsters.guardians.monsters.Team;

/**
 * BattleResult
 * Collects all information about a battle. This includes EXP gained by killing an opponents
 * monster, items, left by opponents and so on.
 *
 * @author Georg Eckert 2017
 */

public class BattleResult
{
    private ArrayMap<AGuardian, Integer> gainedEXP;
    private Array<Item> droppedItems;

    public BattleResult(Team team, Array<Item> droppedItems) {
        this.gainedEXP = new ArrayMap<>();
        for(AGuardian guardian : team.values()) {
            gainedEXP.put(guardian,0);
        }
        this.droppedItems = droppedItems;

    }

    public void gainEXP(AGuardian guardian, int EXP) {
        gainedEXP.put(guardian, gainedEXP.get(guardian) + EXP);
    }

    public boolean applyGainedEXP(AGuardian guardian) {
        boolean levelUp = guardian.getIndividualStatistics().earnEXP(gainedEXP.get(guardian));
        return levelUp;
    }

    /**
     * Applies the earned EXP to all monsters and returns an array of all monsters that reached the
     * next level.
     * @return  {@link Array} of {@link Guardian}s that reached a new level
     */
    public Array<AGuardian> applyGainedEXPtoAll() {
        Array<AGuardian> leveledUpMonsters = new Array<>();
        for(AGuardian guardian : gainedEXP.keys()) {
            boolean lvlUp = applyGainedEXP(guardian);
            if(lvlUp) {
                leveledUpMonsters.add(guardian);
            }
        }
        return leveledUpMonsters;
    }

    public int getGainedEXP(AGuardian guardian) {
        return gainedEXP.get(guardian);
    }
}

File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/battle/BattleResult.kt
package de.limbusdev.guardianmonsters.guardians.battle

import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.ArrayMap

import de.limbusdev.guardianmonsters.guardians.items.Item
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian
import de.limbusdev.guardianmonsters.guardians.monsters.Guardian
import de.limbusdev.guardianmonsters.guardians.monsters.IndividualStatistics
import de.limbusdev.guardianmonsters.guardians.monsters.Team
import de.limbusdev.utils.extensions.set

/**
 * BattleResult
 * Collects all information about a battle. This includes EXP gained by killing an opponents
 * monster, items, left by opponents and so on.
 *
 * @author Georg Eckert 2017
 */
class BattleResult(team: Team, droppedItems: Array<Item>)
{
    // .................................................................................. Properties
    private val gainedEXP    : ArrayMap<AGuardian, Int> = ArrayMap()
    private val droppedItems : Array<Item>

    
    // ................................................................................ Constructors
    init
    {
        team.values().forEach { guardian -> gainedEXP[guardian] = 0 }
        this.droppedItems = droppedItems
    }


    // ..................................................................................... Methods
    fun gainEXP(guardian: AGuardian, EXP: Int)
    {
        gainedEXP[guardian] = gainedEXP[guardian] + EXP
    }

    private fun applyGainedEXP(guardian: AGuardian): Boolean
    {
        val indiStats = guardian.individualStatistics
        indiStats.earnEXP(gainedEXP[guardian])
        return indiStats.didLevelUp()
    }

    /**
     * Applies the earned EXP to all monsters and returns an array of all monsters that reached the
     * next level.
     * @return  [Array] of [Guardian]s that reached a new level
     */
    fun applyGainedEXPtoAll(): Array<AGuardian>
    {
        val leveledUpMonsters = Array<AGuardian>()

        for(guardian in gainedEXP.keys())
        {
            val didLvlUp = applyGainedEXP(guardian)
            if(didLvlUp) { leveledUpMonsters.add(guardian) }
        }

        return leveledUpMonsters
    }

    fun getGainedEXP(guardian: AGuardian): Int = gainedEXP[guardian]
}
